src/Explorer/Tree/ResourceTreeAdapter.tsx

```diff
@@ -351,6 +357,80 @@ export class ResourceTreeAdapter implements ReactAdapter {
     };
   }
 
+  private buildSchemaNode(collection: ViewModels.Collection): TreeNode {
+    if (collection.analyticalStorageTtl() == undefined) {
+      return null;
+    }
+
+    if (
+      collection.schema == undefined ||
+      collection.schema == null ||
+      collection.schema.fields == undefined ||
+      collection.schema.fields == null
+    ) {
+      return null;
+    }
+
+    return {
+      label: "Schema",
+      children: this.getSchemaNodes(collection.schema.fields),
+      onClick: () => {
+        collection.selectedSubnodeKind(ViewModels.CollectionTabKind.Schema);
+        this.container.tabsManager.refreshActiveTab(
+          (tab: TabsBase) => tab.collection && tab.collection.rid === collection.rid
+        );
+      }
+    };
+  }
+
+  private getSchemaNodes(fields: DataModels.IDataField[]): TreeNode[] {
+    let schema: any = {};
+
+    //unflatten
+    fields.forEach((field: DataModels.IDataField, fieldIndex: number) => {
+      const path: string[] = field.path.split(".");
+      const fieldProperties = [field.dataType.name, `HasNulls: ${field.hasNulls}`];
+      let current: any = {};
+      path.forEach((name: string, pathIndex: number) => {
+        if (pathIndex == 0) {
+          if (schema[name] == undefined) {
+            if (pathIndex == path.length - 1) {
+              schema[name] = fieldProperties;
+            } else {
+              schema[name] = {};
+            }
+          }
+          current = schema[name];
+        } else {
+          if (current[name] == undefined) {
+            if (pathIndex == path.length - 1) {
+              current[name] = fieldProperties;
+            } else {
+              current[name] = {};
+            }
+          }
+          current = current[name];
+        }
+      });
+    });
+
+    function traverse(obj: any): TreeNode[] {
```

Comment:
Prefer the fat arrow syntax: `const traverse = (obj: any): TreeNode[] => {}`

Reply:
Consider moving this out of the class or convert to a static function to enable unit testing.

Reply:
Integration test added to ResourceTreeAdapter.test.tsx which covers all within getSchemaNodes including traverse. This is just a POC and depending on what sticks and what does not stick, I refactor afterward when the dust settles to make it unit test friendly.

Reply:
SGTM 👍